feat: snapshot link and IPNS path copy options

jessicaschilling opened the pull request:
Closes https://github.com/ipfs-shipyard/ipfs-companion/issues/740.

This PR moves @deedlefake's work originally in https://github.com/ipfs-shipyard/ipfs-companion/pull/745 to account for some significant changes to the codebase since then.

- [x] Adds "Copy Snapshot Link" menu item that generates immutable permalink in the form of `https://bafy..foo.ipfs.dweb.link/wiki/`
- [x] Adds "Copy Content Path Snapshot" menu item that generates path in the form of ` /ipfs/bafy..foo/wiki/`
- [x] Renames existing "Copy IPFS Content Path" menu item to "Copy IPNS Content Path"
- [x] Clean up hover text for all menu items for better clarity/accuracy
- [x] Wire up smaller gray preview text for new menu items
- [x] Move tools options (import, go to node) into header area for semantic clarity (they apply to all of Companion, not just current tab)
- [x] Check height in Firefox and amend if needed
- [x] Test: make sure things do/don't display as expected on different kinds of sites
     - [x] Local page: Companion welcome page
     - [x] Ordinary web2 page: https://http.cat/
     - [x] Gateway page: https://ipfs.io/ipfs/QmSnuWmxptJZdLJpKRarxBMS2Ju2oANVrgbr2xWbie9b2D
     - [x] DNSLink page: https://docs.ipfs.io/

### Screenshots

Local page
![image](https://user-images.githubusercontent.com/1507828/97764227-8bc07680-1ad3-11eb-9d21-0a9e5e8becef.png)

🚨 🚨 **Ordinary web2 page (this ain't right!):**
![image](https://user-images.githubusercontent.com/1507828/97764434-4b152d00-1ad4-11eb-9eda-4d2325aaaa25.png)

Gateway page:
![image](https://user-images.githubusercontent.com/1507828/97764290-bca0ab80-1ad3-11eb-9f57-baf24caa52b3.png)

DNSLink page:
![image](https://user-images.githubusercontent.com/1507828/97764336-db9f3d80-1ad3-11eb-9bba-25ca6f049002.png)

lidel reviewed (commented):
Quick notes on cosmetics:

- I am  worried about information overload  on DNSLink/IPNS websites
  - is "permalink" clear enough for regular users? (i have no idea - non-native speaker, so feel free to ignore below notes) 
    would users know the difference between "shareable link" and "permalink"? 
    - we do, but that is because we know that IPFS is immutable and IPNS is mutable and can provide updates over time. 
    - on the old internet "permalink" is just "a link that is promised to work in the future", but here we have a perfect snapshot: is there a better language we could use? would "snapshot link" or "revision link"  be bit more explicit? The label should highlight that the link will point to "specific version frozen in time"
- We should copy `https://dweb.link/ipfs/{cid}` as its is the canonical URL
  - it will redirect to proper subdomain: https://docs.ipfs.io/how-to/address-ipfs-on-web/#automatic-leverage-the-gateway-in-go-ipfs
  - makes it easy for people to just swap the hostname or remove hostname, and they get "content path"


ps. some time ago I've added `npm run fix:lint:standard` which should fix most of linting automatically. perhaps we should just run it as pre-commit hook, so people don't need to worry about fixing it manually?

jessicaschilling commented:
@lidel, a few notes on your comment:
1. We're thinking in similar lines about the menu items and helptext. I made some revisions to both in https://github.com/ipfs-shipyard/ipfs-companion/pull/937/commits/6925a743bddc24be974c674998f09434df44fee2 but I think it still needs some work; see updated screenshot above, plus commit info, for latest revisions.
2. Also moved the "Pin IPFS Resources" toggle up so it wasn't at the bottom of a long list of "Copy" items. However, I'm thinking we can still optimize this menu overall so the content is less overwhelming - will draw up some thoughts in future commits.
3. Added `npm run fix:lint:standard` as a pre-commit hook to our existing Husky setup (unless you'd rather handle differently?)
4. I can't figure out how to get hint text for "Copy Shareable Snapshot" and "Copy Content Path Snapshot" (see screenshot above) to display ... do you mind any advice when you get the chance?
5. Regarding copying links as `https://dweb.link/ipfs/{cid}` rather than `https://ipfs.io/ipfs/{cid}` -- would this just be a matter of using the default public subdomain gateway from prefs instead?

lidel reviewed (changes requested):
This is great! 

1. I made some comments on specific menu items, but overall looks great! Adding `Import` / `My Node` buttons made a huge  difference! :heart: 

2. iirc we moved pinning to the bottom because people rarely use it, and copy actions are much more useful – I'd remove changes to pinning item from the scope of this PR (rationale in comment inline)

3. The CI failures are only for `pull_request` tests – those simulate "how master branch would look like if we merge this PR" and are caused by updated standardjs 15.x rules that landed in master when #938 got merged – rebasing this PR should help (ignore it for now)

4. `resolveToPermalink` and `resolveToImmutableIpfsPath` perform some expensive async operations and both return a promise. I've added missing `await` in [`ea9128f` (#937)](https://github.com/ipfs-shipyard/ipfs-companion/pull/937/commits/ea9128fd0d2df890d016d48ccd0ea874988c2145) –  should be fine now
  (this type of bugs will be removed when we add typescript during rewrite required for #666)

5. It also includes cleanup of historical code which we no longer need, which I just realized makes this a bigger task. I'd say out of scope of this PR – created https://github.com/ipfs-shipyard/ipfs-companion/issues/939 to track this separately.

jessicaschilling commented on add-on/_locales/en/messages.json:
```diff
@@ -116,9 +124,17 @@
     "description": "A menu item tooltip in Browser Action pop-up (panel_copyCurrentPublicGwUrlTooltip)"
   },
+  "panel_copyCurrentPermalink": {
+    "message": "Copy Shareable Snapshot Link",
```
Nice rationale - I kept waffling on these titles and you went with my plan B 😊

jessicaschilling commented on add-on/src/popup/browser-action/context-actions.js:
```diff
@@ -56,16 +60,35 @@ function contextActions ({
+    switchValue: (isPinned || isPinning) && !isUnPinning,
+    onClick: isPinned ? onUnPin : onPin
+  })}
+  ${isRedirectContext ? navItem({
```
Thanks! Fixed in https://github.com/ipfs-shipyard/ipfs-companion/pull/937/commits/d02fb7fd65a4ff81b54ce0a436f6b382c0041f2e

jessicaschilling commented:
@lidel -- All items above taken care of, but two remaining concerns:

1. Getting an "offline" state for ordinary http sites (see updated screenshots in first post)
2. Also getting rapidly cycling "IPFS API is (Online/Offline)" browser notifications, but can't seem to find any actions in the UI that are triggering these

Do you mind having a look when you are able? Thank you!

lidel commented:
@jessicaschilling landed additional optimizations in b369846 and 6cc79cf  – give it a spin in a spare moment :)

Changes:

- URL-to-Path resolution now runs asynchronously and result is stored in the cache, then value is read from the cache, which makes UI snappy and not blocked by slow HTTP API.
  - we can now reuse cached values during copying, but I've decided it's out of scope (can be a separate PR later)
- Tested labels a bit and applied some tweaks:
  - Swapped logic a bit  to ensure  "Copy Shareable Link" is shown on both `/ipns/` and `/ipfs/` 
    - "Snapshot" makes sense only in the context of `/ipns/` 
  - Changed path labels back to `Copy IPFS/IPNS Path` 
	- I believe the path form is more useful for technical user: regular user will care about "shareable/snapshot links", however technical user will understand content paths labeled as IPFS vs IPNS better.
  - `/ipfs/` looks like this:
    >  ![ipfs-2020-11-07--03-19-23](https://user-images.githubusercontent.com/157609/98429632-2575be00-20a8-11eb-92ca-0adc8f741b7b.png)
  - `/ipns`/ looks like this:
     > ![2020-11-07--03-19-52](https://user-images.githubusercontent.com/157609/98429635-2b6b9f00-20a8-11eb-98c8-4cf1ee182653.png)
- There was also change to the `pin.ls` API, which caused the "offline" state sometimes (instead of checking specific CID it listed ALL pinned blocks, which could take some time, during which UI was dead, just like you described).

jessicaschilling commented:
@lidel LGTM! Agree on "IPNS Path"/"IPFS Path" in the menu items - hover labels disambiguate this if need be. Please feel free to merge ... thank you so much!

lidel commented:
![2020-11-10--00-40-01](https://user-images.githubusercontent.com/157609/98609326-5266e280-22ed-11eb-9076-4b216c1417bc.png)
